### PR TITLE
[DependencyInjection] introduce feature switch to disable S.R.E

### PR DESCRIPTION
**Pull request by jonathanpeppers**
When recording a new AOT profile for .NET MAUI apps running on Android, we noticed that System.Reflection.Emit work was being done on a background thread. The call seen in `dotnet-trace` output:

    11.32ms microsoft.extensions.dependencyinjection!Microsoft.Extensions.DependencyInjection.ServiceLookup.ILEmitResolverBuilder.GenerateMethodBody(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite,System.Reflection.Emit.ILGenerator)

.NET Android apps are unique in that there is a JIT, `RuntimeFeature.IsDynamicCodeCompiled` is true, System.Reflection.Emit is possible -- S.R.E is however, not great for startup performance.

Starting threads on Android during startup can also be slow, as Android will commonly put all but a single core to sleep for battery saving purposes. We try to avoid starting threads on startup for "hello world" applications on Android.

To solve this for now, introduce a new feature flag:

    Microsoft.Extensions.DependencyInjection.DisableDynamicEngine

Which, we will provide a value in either the Android or .NET MAUI optional workload via an MSBuild property. To test, I put this in my app's `.csproj` file:

    <RuntimeHostConfigurationOption Include="Microsoft.Extensions.DependencyInjection.DisableDynamicEngine"
                                    Condition="'$(DisableDynamicEngine)' != ''"
                                    Value="$(DisableDynamicEngine)"
                                    Trim="true" />

Customers *could* opt to change this flag, but we don't think it will particularly useful. An example of services realized by .NET MAUI at startup, via some logging added:

    08-25 13:21:55.647 16530 16530 I DOTNET  : RealizeService called: System.Collections.Generic.IEnumerable`1[Microsoft.Maui.Hosting.IMauiInitializeService]
    08-25 13:21:55.664 16530 16530 I DOTNET  : RealizeService called: System.Collections.Generic.IEnumerable`1[Microsoft.Maui.Hosting.IMauiInitializeScopedService]
    08-25 13:21:55.665 16530 16530 I DOTNET  : RealizeService called: Microsoft.Maui.Dispatching.IDispatcher
    08-25 13:21:55.668 16530 16530 I DOTNET  : RealizeService called: System.Collections.Generic.IEnumerable`1[Microsoft.Maui.LifecycleEvents.LifecycleEventRegistration]
    08-25 13:21:56.057 16530 16530 I DOTNET  : RealizeService called: System.Collections.Generic.IEnumerable`1[Microsoft.Maui.Hosting.HandlerMauiAppBuilderExtensions+HandlerRegistration]
    08-25 13:21:56.115 16530 16530 I DOTNET  : RealizeService called: Microsoft.Extensions.DependencyInjection.IServiceScopeFactory
    08-25 13:21:56.670 16530 16530 I DOTNET  : RealizeService called: Microsoft.Maui.Controls.HideSoftInputOnTappedChangedManager
    08-25 13:21:56.712 16530 16530 I DOTNET  : RealizeService called: System.Collections.Generic.IEnumerable`1[Microsoft.Maui.Hosting.ImageSourcesMauiAppBuilderExtensions+ImageSourceRegistration]
    08-25 13:21:57.700 16530 16530 I DOTNET  : RealizeService using S.R.E: Microsoft.Maui.Controls.HideSoftInputOnTappedChangedManager

`HideSoftInputOnTappedChangedManager` would be realized once per page, which would not be a huge payoff to use S.R.E for. So the only way the S.R.E codepath could be useful on Android would be if the customer is registering lots of services themselves. They might be better off just using `new()` in that case?

An example of the startup time Android reports with the new flag on/off:

    DisableDynamicEngine=false
    08-25 14:31:37.462  2090  2330 I ActivityTaskManager: Displayed com.companyname.testmaui/crc643c09abdeec717b83.MainActivity: +733ms
    08-25 14:31:39.394  2090  2330 I ActivityTaskManager: Displayed com.companyname.testmaui/crc643c09abdeec717b83.MainActivity: +737ms
    08-25 14:31:41.326  2090  2330 I ActivityTaskManager: Displayed com.companyname.testmaui/crc643c09abdeec717b83.MainActivity: +730ms
    DisableDynamicEngine=true
    08-25 14:32:20.233  2090  2330 I ActivityTaskManager: Displayed com.companyname.testmaui/crc643c09abdeec717b83.MainActivity: +724ms
    08-25 14:32:22.137  2090  2330 I ActivityTaskManager: Displayed com.companyname.testmaui/crc643c09abdeec717b83.MainActivity: +727ms
    08-25 14:32:24.042  2090  2330 I ActivityTaskManager: Displayed com.companyname.testmaui/crc643c09abdeec717b83.MainActivity: +716ms

This was a `dotnet new maui` project, using dotnet/maui/main on a Pixel 5 device.

### CODE REVIEW

File: docs/workflow/trimming/feature-switches.md [hunk 1]
```diff
@@ -27,6 +27,7 @@ configurations but their defaults might vary as any SDK can set the defaults dif
 | MetadataUpdaterSupport | System.Reflection.Metadata.MetadataUpdater.IsSupported | Metadata update related code to be trimmed when set to false |
 | _EnableConsumingManagedCodeFromNativeHosting | System.Runtime.InteropServices.EnableConsumingManagedCodeFromNativeHosting | Getting a managed function from native hosting is disabled when set to false and related functionality can be trimmed. |
 | VerifyDependencyInjectionOpenGenericServiceTrimmability | Microsoft.Extensions.DependencyInjection.VerifyOpenGenericServiceTrimmability | When set to true, DependencyInjection will verify trimming annotations applied to open generic services are correct |
+| DisableDynamicEngine | Microsoft.Extensions.DependencyInjection.DisableDynamicEngine | When set to true, DependencyInjection will avoid using System.Reflection.Emit when realizing services |
```
Comment (eerhardt) [inline docs/workflow/trimming/feature-switches.md:30]: `DisableDynamicEngine` is probably too generic of a name. It should be more specific to DI.

```suggestion
| DisableDependencyInjectionDynamicEngine | Microsoft.Extensions.DependencyInjection.DisableDynamicEngine | When set to true, DependencyInjection will avoid using System.Reflection.Emit when realizing services |
```

**Comment by steveharter** [comment]
Note that adding such a feature switch was tentatively on the plan when S.R.E was enabled. See https://github.com/dotnet/runtime/issues/66153 and the section:

_If perf is not acceptable for all cases, add a new feature switch that will use LINQ instead of reflection. It would be off by default in WASM, but could be turned on if LINQ is already used elsewhere in the app, for example. Update we need to verify against actual ASP.NET benchmarks, but perf is likely acceptable given the trade-off in Blazor is acceptable CPU perf tradeoff vs. having the large LINQ assembly._

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs [hunk 2]
```diff
@@ -246,7 +249,7 @@ private ServiceProviderEngine GetEngine()
 #if NETFRAMEWORK || NETSTANDARD2_0
             engine = CreateDynamicEngine();
 #else
-            if (RuntimeFeature.IsDynamicCodeCompiled)
+            if (RuntimeFeature.IsDynamicCodeCompiled && !DisableDynamicEngine)
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs:252]: For other reviewers: the check for `IsDynamicCodeCompiled` here was added prior to v8.

**Comment by steveharter** [comment]
Note the related emit+Android issue where emit is also to be avoided for reflection invoke: https://github.com/dotnet/runtime/issues/83893 which uses `Switch.System.Reflection.ForceInterpretedInvoke`.

Should the changes there that use `ForceInterpretedInvoke` be changed to use the new switch added here?

**Comment by jonathanpeppers** [comment]
My plan was to add changes to the `Android` workload in a similar way as: https://github.com/xamarin/xamarin-android/pull/7972

```xml
<DisableDependencyInjectionDynamicEngine Condition="'$(DisableDependencyInjectionDynamicEngine)' == ''">true</DisableDependencyInjectionDynamicEngine>
```
I wouldn't expect many of our customers to know about these or try to use them.

I can unify these both to use a single MSBuild property, though -- thoughts on a good name? `$(DisableSystemReflectionEmitOptimizations)`?

File: docs/workflow/trimming/feature-switches.md [hunk 3]
```diff
@@ -27,6 +27,7 @@ configurations but their defaults might vary as any SDK can set the defaults dif
 | MetadataUpdaterSupport | System.Reflection.Metadata.MetadataUpdater.IsSupported | Metadata update related code to be trimmed when set to false |
 | _EnableConsumingManagedCodeFromNativeHosting | System.Runtime.InteropServices.EnableConsumingManagedCodeFromNativeHosting | Getting a managed function from native hosting is disabled when set to false and related functionality can be trimmed. |
 | VerifyDependencyInjectionOpenGenericServiceTrimmability | Microsoft.Extensions.DependencyInjection.VerifyOpenGenericServiceTrimmability | When set to true, DependencyInjection will verify trimming annotations applied to open generic services are correct |
+| DisableDependencyInjectionDynamicEngine | Microsoft.Extensions.DependencyInjection.DisableDynamicEngine | When set to true, DependencyInjection will avoid using System.Reflection.Emit when realizing services |
```
Comment (steveharter) [inline docs/workflow/trimming/feature-switches.md:30]: I wonder if a more universal "AvoidEmitForPerformance" switch makes sense. That would also be checked in the couple places where the undocumented `ForceInterpretedInvoke` is used (which was originally added for testing).

Comment (jonathanpeppers) [reply docs/workflow/trimming/feature-switches.md:30]: Yes, I think it makes sense to rework https://github.com/xamarin/xamarin-android/pull/7972 so it sets both feature switches on a name similar to `$(AvoidEmitForPerformance)`.

We would try to ship this in .NET 8 RC 2 if possible.

**Review by steveharter — APPROVED** [review]
Per offline discussion, we may want to remove the .md file changes since this switch is considered internal and may be replaced by "AvoidEmitForPerformance" (or a better name) in the future where that new switch would also by used by Reflection and DependencyInjection for emit cases where today they use IsDynamicCodeSupported.

**Pull request merged**